satellite/gc/service.go

```diff
@@ -0,0 +1,171 @@
+	// value for InitialPieces currently based on average pieces per node
+	InitialPieces     int     `help:"the initial number of pieces expected for a storage node to have, used for creating a filter" releaseDefault:"400000" devDefault:"10"`
+	FalsePositiveRate float64 `help:"the false positive rate used for creating a filter" releaseDefault:"0.1" devDefault:"0.1"`
+	ConcurrentSends   int     `help:"the number of nodes to concurrently send bloom filters to" releaseDefault:"1" devDefault:"1"`
+}
+
+// Service implements the garbage collection service
+type Service struct {
+	log          *zap.Logger
+	Loop         *sync2.Cycle
+	metainfoloop *metainfo.Loop
+	transport    transport.Client
+	overlay      overlay.DB
+	config       Config
+
+	lastPieceCounts map[storj.NodeID]int
+
+	pieceCountsMutex sync.Mutex
+	pieceCounts      map[storj.NodeID]int
+}
+
+// RetainInfo contains info needed for a storage node to retain important data and delete garbage data
+type RetainInfo struct {
+	Filter       *bloomfilter.Filter
+	CreationDate time.Time
+	count        int
+}
+
+// NewService creates a new instance of the gc service
+func NewService(log *zap.Logger, transport transport.Client, overlay overlay.DB, loop *metainfo.Loop, config Config) *Service {
+	// TODO retrieve piece counts from overlay (when there is a column for them)
+	// var lastPieceCounts atomic.Value
+	// lastPieceCounts.Store(map[storj.NodeID]int{})
+	lastPieceCounts := make(map[storj.NodeID]int)
+
+	return &Service{
+		log:              log,
+		Loop:             sync2.NewCycle(config.Interval),
+		metainfoloop:     loop,
+		transport:        transport,
+		overlay:          overlay,
+		lastPieceCounts:  lastPieceCounts,
+		config:           config,
+		pieceCountsMutex: sync.Mutex{},
+	}
+}
+
+// Run starts the gc loop service
+func (service *Service) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	return service.Loop.Run(ctx, func(ctx context.Context) error {
+		obs := NewObserver(service.log.Named("gc observer"), service.lastPieceCounts, service.config)
+
+		err := service.metainfoloop.Join(ctx, obs)
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		err = service.Send(ctx, obs)
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		return nil
+	})
+}
+
+// Send sends the piece retain requests to all storage nodes
+func (service *Service) Send(ctx context.Context, obs *Observer) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	service.pieceCounts = make(map[storj.NodeID]int)
+
+	limiter := sync2.NewLimiter(service.config.ConcurrentSends)
+	for id, retainInfo := range obs.retainInfos {
+		func(id storj.NodeID, retainInfo *RetainInfo) {
+			limiter.Go(ctx, func() {
+				err := service.sendRetainRequest(ctx, id, retainInfo)
+				if err != nil {
+					service.log.Error("error sending retain info", zap.Error(err))
+				}
+			})
+		}(id, retainInfo)
+	}
+	limiter.Wait()
+
+	service.lastPieceCounts = service.pieceCounts
```

Comment:
Why don't we do this before dealing with the limiter?

Reply:
the order shouldn't really matter. We just need update it after `metainfoLoop.Join`

Reply:
updated to send bloom filters last